### PR TITLE
Added error handling for a ResourceNotFound::Deployment

### PR DESCRIPTION
**Pull request by silvestrpredko**

### CODE REVIEW

File: lumeo_api_client/src/error.rs [hunk 1]
```diff
@@ -2,31 +2,55 @@ use std::fmt;
...
+    value
+        .as_object()
+        .and_then(|obj| obj.values().next())
+        .and_then(|resource| resource.as_str())
+        .and_then(|it| {
+            it.parse::<ResourceNotFound>()
+                .unwrap_or_else(|_| ResourceNotFound::Other(it.to_owned()))
+                .into()
+        })
+}
```
Comment (FSMaxB) [inline lumeo_api_client/src/error.rs:46]:
```suggestion
impl ResourceNotFound {
    fn from_context(mut context: serde_json::Value) -> Option<Self> {
        let resource = match context.as_object_mut()?
            .remove("resource")? {
            serde_json::Value::String(resource) => resource,
            _ => return None,
        };

        let resource_not_found = resource.parse::<ResourceNotFound>()
            .unwrap_or_else(|_| ResourceNotFound::Other(resource));
        Some(resource_not_found)
    }
}
```

You could also use the following if you don't mind the string copy of the `to_owned`:

```rust
impl ResourceNotFound {
    fn from_context(context: &serde_json::Value) -> Option<Self> {
        let resource = context.as_object()?.get("resource")?.as_str()?;

        let resource_not_found = resource.parse::<ResourceNotFound>()
            .unwrap_or_else(|_| ResourceNotFound::Other(resource.to_owned()));
        Some(resource_not_found)
    }
}

```

**Review by FSMaxB — APPROVED** [review]

File: lumeo_api_client/src/error.rs [hunk 2]
```diff
@@ -35,9 +59,17 @@ impl<'de> Deserialize<'de> for ApiError {
...
     {
         let resp = ApiServerResponse::deserialize(deserializer)?;
-        Ok(match resp.code.parse::<ApiError>() {
-            Ok(api_err) => api_err,
-            Err(_) => ApiError::Other { code: resp.code, message: resp.message },
+
+        Ok(match &resp.code {
+            err_name if err_name == ApiError::VARIANTS[0] => ApiError::GatewayDeleted,
+            err_name if err_name == ApiError::VARIANTS[1] => ApiError::InvalidCredentials,
+            err_name if err_name == ApiError::VARIANTS[2] => {
```
[the remark this answers is not in the document]
Comment (jplatte) [reply lumeo_api_client/src/error.rs:66]: The `AsStaticRef` trait has been deprecated, but the `AsRefStr` derive hasn't. We should use it.

File: lumeo_api_client/src/error.rs [hunk 3]
```diff
@@ -155,21 +187,30 @@ impl<T> ResultExt<T> for Result<T, reqwest::Error> {
 mod tests {
     use super::*;
 
+    #[test]
+    fn resource_not_found() {
+        let resp = ApiServerResponse {
+            code: "resource-not-found".to_owned(),
+            context: serde_json::json!({"resource": "deployment"}).into(),
```
Comment (jplatte) [inline lumeo_api_client/src/error.rs:194]:
```suggestion
            context: serde_json::json!({ "resource": "deployment" }).into(),
```

File: lumeo_api_client/src/error.rs [hunk 4]
```diff
@@ -58,9 +59,17 @@ impl<'de> Deserialize<'de> for ApiError {
         D: serde::Deserializer<'de>,
     {
         let resp = ApiServerResponse::deserialize(deserializer)?;
```
Comment (DmitrySamoylov) [inline lumeo_api_client/src/error.rs:61]: The below part can be simplified a bit if `resp` is destructured here:
```rust
        let ApiServerResponse { code, message, context } =
            ApiServerResponse::deserialize(deserializer)?;
```

File: lumeo_api_client/src/error.rs [hunk 5]
```diff
@@ -2,42 +2,83 @@ use std::fmt;
...
         D: serde::Deserializer<'de>,
     {
-        let resp = ApiServerResponse::deserialize(deserializer)?;
-        Ok(match resp.code.parse::<ApiError>() {
-            Ok(api_err) => api_err,
-            Err(_) => ApiError::Other { code: resp.code, message: resp.message },
+        let ApiServerResponse { code, message, context } =
+            ApiServerResponse::deserialize(deserializer)?;
+
+        Ok(match &code {
```
Comment (jplatte) [inline lumeo_api_client/src/error.rs:71]: Given that there is no structural matching happening at all, I think this `match` would be easier to read as an `if` / `else if` / `else` expression.

Comment (silvestrpredko) [reply lumeo_api_client/src/error.rs:71]: Make sense

**Review by jplatte — APPROVED** [review]

**Pull request merged**